docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,158 @@
+# Use SPIFFE/SPIRE for On-demand Secret Store Token Generation
+
+## Status
+
+** DRAFT **
+
+## Context
+
+In security-enabled EdgeX,
+there is a component called `security-secretstore-setup`
+that seeds authentication tokens for Hashicorp Vault--EdgeX's secret store--into
+directories reserved for each EdgeX microservice.
+The implementation is provided by a sub-component, `security-file-token-provider`,
+that works off of a static configuration file and an environment variable
+that lists the services that require tokens
+and the policy to be attached to those tokens.
+
+The current solution has some problematic aspects:
+
+* These tokens have an initial TTL of one hour (1h)
+  and become invalid if not used and renewed within that time period.
+  In practical terms,
+  this means that the EdgeX framework must be shut down and restarted completely
+  in order to change the list of running services.
+  It is not possible to delay the start of EdgeX services until a later time,
+  as they will not be able to connect to the EdgeX secret store
+  to obtain required secrets.
+
+* Transmission of the authentication token requires one or more shared file systems
+  between the service and `security-secretstore-setup`.
+  In the Docker implementation,
+  this shared file system is implemented by host-based local storage.
+  Similar functionality is near-impossible in a Kubernetes environment
+  that supports only RWO volumes:
+  the volume containing the secret can be mounted to only a single pod at a time.
+  It is simply easier to run without security enabled in Kubernetes.
+
+* The current approach cannot support distributed services
+  without an underlying distributed file system to distribute tokens,
+  such as GlusterFS, running across the participating nodes.
+
+## Decision
+
+EdgeX will create a new service, `security-spiffe-token-provider`.
+This service will be a mutual-auth TLS service that exchanges
+a [SPIFFE](https://spiffe.io/) X.509 SVID for a secret store token.
+
+The `go-mod-secrets` module will be modified enable a new mode
+whereby a secret store token is obtained by:
+
+1. Obtaining an X.509 SVID by contacting a local SPIFFE agent's
+   workload API on a local Unix domain socket.
+2. Connecting to the `security-spiffe-token-provider` service
+   and using the X.509 SVID and request a secret store token.
+
+The SPIFFE authentication mode will be the default authentication
+mode for non-core EdgeX services and for remote EdgeX services.
```

Comment:
Should we call out these services explicitly? Namely support-* and app-service-configurable?

Reply:
Changed to "The SPIFFE authentication mode will be an opt-in feature."